src/main/java/com/pusher/client/channel/impl/PrivateEncryptedChannelImpl.java

```diff
@@ -22,6 +25,23 @@
     private SecretBoxOpenerFactory secretBoxOpenerFactory;
     private SecretBoxOpener secretBoxOpener;
 
+    // For not hanging on to shared secret past the Pusher.disconnect() call,
+    // i.e. when not necessary. Pusher.connect(...) call will trigger re-subscribe
+    // and hence re-authenticate which creates a new secretBoxOpener.
+    private ConnectionEventListener disposeSecretBoxOpenerOnDisconnectedListener =
+            new ConnectionEventListener() {
+
+        @Override
+        public void onConnectionStateChange(ConnectionStateChange change) {
+            disposeSecretBoxOpener();
```

Comment:
Won't this get called a lot of times? e.g. you go through CONNECTING to CONNECTED - would this not be better if it was only if the connect state was DISCONNECTING?

Reply:
It's a bit... "disconnected", but you'll find your answer here: https://github.com/pusher/pusher-websocket-java/pull/247/files#diff-48ff2981ea97376b5f1cedec4b59e059R114